### ISSUE TITLE
Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.

### ISSUE THREAD

**Issue by chobo2** [aspnet/AspNetCore#13259]
Hi
I am getting this error

```
   Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
       at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
       at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
       at Mapster.TypeAdapterConfig.Remove(TypeTuple key)
       at Mapster.TypeAdapterConfig.NewConfig[TSource,TDestination]()
       at InventoryService.GetCategories(Int32 inventoryCenterId)
       at Controllers.InventoriesController.GetCategories(Int32 centerId) 

```

The code it is pointing to is this

```
  TypeAdapterConfig<Category, CategoryDto>
               .NewConfig()
               .IgnoreIf((src, dest) => string.IsNullOrEmpty(src.MetaTitle), dest => dest.MetaTitle)
              .IgnoreIf((src, dest) => string.IsNullOrEmpty(src.MetaDescription), dest => dest.MetaDescription);

            var categories = dbContext.Centers.Where(x => x.Id == CenterId).SelectMany(x => x.Categories).OrderBy(x => x.Name).ToList();
            return categories.Adapt<List<CategoryDto>>();
```
Starting and stopping IIS fixes this for a certain amount of time and then it comes back.

I found these related issues

https://github.com/aspnet/AspNetCore/issues/4970

https://github.com/IdentityServer/IdentityServer4/issues/2453

but I am using 2.2 for Aspnet Core and Ef Core 2.2.4 and they where using older versions.

edit

I ran on the the server this opersation

 (dir (Get-Command dotnet).Path.Replace('dotnet.exe', 'shared\Microsoft.NETCore.App')).Name

and I am getting version 2.2.5

**Comment by rynowak** [comment]
Is `Mapster` a library that you're using? Do they have documentation about how to use it correctly in ASP.NET Core?

This kind of error means what it says. Request processing in servers runs in parallel - your code is running something that's not thread-safe on multiple threads at a time.

**Comment by chobo2** [comment]
> Is `Mapster` a library that you're using? Do they have documentation about how to use it correctly in ASP.NET Core?
> 
> This kind of error means what it says. Request processing in servers runs in parallel - your code is running something that's not thread-safe on multiple threads at a time.

Hey

Yes Mapster is a library: https://github.com/MapsterMapper/Mapster  

I think I am using it properly. I don't see anything specific pertaining asp.net core, though a book on asp.net mvc recommended to use it.

**Comment by rynowak** [comment]
The best hint I could give you then is probably this:

```txt
  at Mapster.TypeAdapterConfig.NewConfig[TSource,TDestination]()
       at InventoryService.GetCategories(Int32 inventoryCenterId)
       at Controllers.InventoriesController.GetCategories(Int32 centerId)
```

This call stack implies that calling into `InventoriesController` will end up initializing your mappings, and that means that it can probably happen concurrently. You'll want to find a way to initialize your mappings once (probably during startup).

**Comment by chobo2** [comment]
> The best hint I could give you then is probably this:
> 
> ```
>   at Mapster.TypeAdapterConfig.NewConfig[TSource,TDestination]()
>        at InventoryService.GetCategories(Int32 inventoryCenterId)
>        at Controllers.InventoriesController.GetCategories(Int32 centerId)
> ```
> 
> This call stack implies that calling into `InventoriesController` will end up initializing your mappings, and that means that it can probably happen concurrently. You'll want to find a way to initialize your mappings once (probably during startup).

Hmm, why would it just happen randomly and not always then?

Here is where I got the code from. Should I try the global settings and put it in up startup?

https://github.com/MapsterMapper/Mapster/wiki/Configuration

**Comment by rynowak** [comment]
> Hmm, why would it just happen randomly and not always then?

It would happen randomly because it relies on two threads both trying to modify that dictionary in the same instant. Even if you're handling a fair bit of traffic, they won't always line up perfectly.

I'm not sure if you need to change what you're doing substantially - have you tried moving the mapping initialization to the constructor of `InventoryService`? Is `InventoryService` a singleton?

**Comment by chobo2** [comment]
> > Hmm, why would it just happen randomly and not always then?
> 
> It would happen randomly because it relies on two threads both trying to modify that dictionary in the same instant. Even if you're handling a fair bit of traffic, they won't always line up perfectly.
> 
> I'm not sure if you need to change what you're doing substantially - have you tried moving the mapping initialization to the constructor of `InventoryService`? Is `InventoryService` a singleton?

I could try moving it to constructor, I tried moving it to startup and it functions but won't be able to tell until it runs for awhile

All my services are "AddTransient"

**Comment by chobo2** [comment]
@rynowak 

I have not seen the problem since moving out that code to startup but also not have hit that same threshold of users so can't put the issue to rest yet.

However I did get this error and not sure if it is related or a different issue(if it is then I will create a new ticket if warranted) 

```
Microsoft.AspNetCore.Server.Kestrel|Connection id “0HLP5AO39QCJI”, Request id “0HLP5AO39QCJI:00000017”: An unhandled exception was thrown by the application. Microsoft.AspNetCore.Connections.ConnectionResetException: An existing connection was forcibly closed by the remote host ---> System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote host

at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitableEventArgs.<GetResult>g__ThrowSocketException|7_0(SocketError e)
at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitableEventArgs.GetResult()
at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.ProcessReceives()
at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.DoReceive()
--- End of inner exception stack trace ---
at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
at System.IO.Pipelines.Pipe.GetReadAsyncResult()
at System.IO.Pipelines.Pipe.DefaultPipeReader.GetResult(Int16 token)
at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1MessageBody.PumpAsync()
at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
at System.IO.Pipelines.Pipe.GetReadAsyncResult()
at System.IO.Pipelines.Pipe.DefaultPipeReader.GetResult(Int16 token)
at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.MessageBody.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
at Microsoft.AspNetCore.WebUtilities.FileBufferingReadStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
at Microsoft.AspNetCore.WebUtilities.StreamHelperExtensions.DrainAsync(Stream stream, ArrayPool`1 bytePool, Nullable`1 limit, CancellationToken cancellationToken)
at Microsoft.AspNetCore.Mvc.Formatters.JsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
at Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinder.BindModelAsync(ModelBindingContext bindingContext)
at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, Object value)
at Microsoft.AspNetCore.Mvc.Internal.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---

at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
at Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)
at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.Invoke(HttpContext httpContext)
at Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware.InvokeCore(HttpContext context)
at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
at Microsoft.AspNetCore.Server.IISIntegration.IISMiddleware.Invoke(HttpContext httpContext)
at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)

```

**Comment by rynowak** [comment]
That looks like a normal disconnect - ie: client disconnected after sending incomplete data. Do you have any reason to believe this is a problem in your app code?

**Comment by chobo2** [comment]
> That looks like a normal disconnect - ie: client disconnected after sending incomplete data. Do you have any reason to believe this is a problem in your app code?

No real reason to believe it would be code issue just one customer has reported seeing some message that said something like "api uploading 0%" but I am still trying to get hold of that customer to do a screen share, so in the mean time I saw this error that I never seen before and thought may they are linked.

Will keep you posted.

**Comment by chobo2** [comment]
@rynowak 

Hey so I met with the customer and watched his screen and I can see that this message directly correlates to him. 

Why would incomplete data happen? Is it because his connection is slow or high latency?

It was very had to trouble shoot as I had to do a screen share with him and his internet connection seemed very bad. I did see in the developer console he was getting alot of "Connection Refused" errors.

I did a speed test and saw he was getting like 2 megabits per second with a ping of 109ms. So not sure if it was connection speed was causing troubles.

**Comment by rynowak** [comment]
You'd see a disconnect I think if there's significant packet loss. It sounds like your user is having some basic network reliability problems. @Tratcher any additional thoughts

**Comment by chobo2** [comment]
> You'd see a disconnect I think if there's significant packet loss. It sounds like your user is having some basic network reliability problems. @Tratcher any additional thoughts

That's what I thought, though he was saying that he can access every other site just fine with no problems but I could not verify this as the screen share was a slide show(probably due to his horrible speeds).

I have asked him to try accessing the sites at different locations to see if that makes a difference but till I hear back I am just trying to drum up ideas on what could be going on as he has been rated a higher value customer so a solution needs to be found.

**Comment by Tratcher** [comment]
If it were a throughput issue you'd have a different log from kestrel saying that the server terminated the connection due to a low data rate. That ping is high, but not prohibitively so. If you run a `ping -t` are you seeing a lot of dropped packets? A wireshark trace would show a lot of retransmit requests.

**Comment by chobo2** [comment]
> If it were a throughput issue you'd have a different log from kestrel saying that the server terminated the connection due to a low data rate. That ping is high, but not prohibitively so. If you run a `ping -t` are you seeing a lot of dropped packets? A wireshark trace would show a lot of retransmit requests.

hmm, well the customer has come back and said he tried at home and it worked great. So he is going to check with his internet provider about his connection.

I howver am still getting this error and I am not sure if this is another customer that is having this problem or what. Is there away to collect the information to figure this out or try to reproduce this error.

I am just not sure how big of an issue this is right now.

**Comment by Tratcher** [comment]
Unfortunately this kind of issue needs to be analyzed at the network layer or from the client end. E.g. a (decrypted) wireshark trace on the client or server during the repro would help explain why the disconnect happened.

**Comment by chobo2** [comment]
> Unfortunately this kind of issue needs to be analyzed at the network layer or from the client end. E.g. a (decrypted) wireshark trace on the client or server during the repro would help explain why the disconnect happened.

so this is purely a network issue and not a code issue and has to be checked at this level?

**Comment by Tratcher** [comment]
Likely. If there were something in the app causing this there would be a preceding log about the server closing the connection for a given reason.

**Comment by chobo2** [comment]
> Likely. If there were something in the app causing this there would be a preceding log about the server closing the connection for a given reason.

ok, I will contact my companies azure guy and see if they have any insight with this. Any other recommendations on next steps?

**Comment by Tratcher** [comment]
No, client disconnects are very normal. They only throw an exception here so you can tell that the request body was incomplete and you don't try to process it.
